pkg/trie/triedb/node.go

```diff
@@ -168,8 +207,204 @@ func newNodeFromEncoded(nodeHash common.Hash, data []byte, storage NodeStorage)
 			children[i] = child
 		}
 
-		return Branch{partialKey: key, children: children, value: NewFromEncoded(value)}, nil
+		return Branch{partialKey: key, children: children, value: NewValueFromEncoded(encoded.PartialKey, value)}, nil
+	default:
+		panic("unreachable")
+	}
+}
+
+type NodeToEncode interface {
+	isNodeToEncode()
+}
+
+type (
+	NewNodeToEncode struct {
+		partialKey []byte
+		value      []byte
+	}
+	TrieNodeToEncode struct {
+		child NodeHandle
+	}
+)
+
+func (NewNodeToEncode) isNodeToEncode()  {}
+func (TrieNodeToEncode) isNodeToEncode() {}
+
+type ChildReference interface {
+	isChildReference()
+	getNodeData() []byte
+}
+
+type (
+	HashChildReference struct {
+		hash common.Hash
+	}
+	InlineChildReference struct {
+		encodedNode []byte
+	}
+)
+
+func (HashChildReference) isChildReference() {}
+func (h HashChildReference) getNodeData() []byte {
+	return h.hash.ToBytes()
+}
+func (InlineChildReference) isChildReference() {}
+func (i InlineChildReference) getNodeData() []byte {
+	return i.encodedNode
+}
+
+type childFunc = func(node NodeToEncode, oslice []byte, oindex *byte) (ChildReference, error)
```

Comment:
I think would be useful to add a comment here describing what function types purpose is

Reply:
I think it's more clear now that I renamed it, but let me know what you think